rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -836,148 +823,282 @@ rmw_ret_t configure_qos_for_security(
+    return RMW_RET_ERROR;
+  }
 
-       Note that "domains_lock" is held all this time. */
-    gcdds.domains[did].domain_handle = dds_get_parent(pp);
+  std::unique_ptr<rmw_guard_condition_t, std::function<void(rmw_guard_condition_t *)>>
+  graph_guard_condition(
+    create_guard_condition(impl.get()),
+    [&](rmw_guard_condition_t * gc) {
+      if (RMW_RET_OK != destroy_guard_condition(gc)) {
+        RMW_SAFE_FWRITE_TO_STDERR(
+          "Failed to destroy guard condition after function: '"
+          RCUTILS_STRINGIFY(__function__) "' failed.\n");
+      }
+    });
+  if (!graph_guard_condition) {
+    return RMW_RET_BAD_ALLOC;
+  }
+
+  impl->common.graph_cache.set_on_change_callback(
+    [guard_condition = graph_guard_condition.get()]() {
+      static_cast<void>(rmw_trigger_guard_condition(guard_condition));
+    });
+
+  impl->common.pub = publisher.get();
+  impl->common.sub = subscription.get();
+  impl->common.graph_guard_condition = graph_guard_condition.get();
+  get_entity_gid(impl->ppant, impl->common.gid);
+  impl->common.graph_cache.add_participant(impl->common.gid, context->options.security_context);
+  impl->common.thread_is_running.store(false);
+  impl->common.listener_thread_gc = nullptr;
+
+  context->impl = impl.get();
+
+  // FIXME: not sure if I could get away with using a bunch of listeners
+  // - Locally published samples shouldn't make it to the reader, so there shouldn't be a deadlock
+  //   caused by the graph cache's mutex already having been locked by (e.g.) rmw_create_node.
+  // - Whatever the graph cache implementation does, it shouldn't involve much more than local state
+  //   updates and triggering a guard condition, and so that should be safe.
+  // - There might be a significant cost to calling the graph cache function to update its internal
+  //   state, and having that occur on the thread receiving data from the network may not be wise
+  // So it may be better to create a thread and a waitset.
+  if ((ret = discovery_thread_start(context)) != RMW_RET_OK) {
+    return ret;
+  }
+
+  graph_guard_condition.release();
+  publisher.release();
+  subscription.release();
+  impl.release();
+  return RMW_RET_OK;
+}
+
+extern "C" rmw_ret_t rmw_shutdown(rmw_context_t * context)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(context, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    context,
+    context->implementation_identifier,
+    eclipse_cyclonedds_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+  // Nothing to do here for now.
+  // This is just the middleware's notification that shutdown was called.
+  return RMW_RET_OK;
+}
+
+extern "C" rmw_ret_t rmw_context_fini(rmw_context_t * context)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(context, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    context,
+    context->implementation_identifier,
+    eclipse_cyclonedds_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+  delete context->impl;
```

Comment:
Good catch! The RMW publisher, subscription and guard condition objects for interfacing wiht the graph cache weren't deleted (the underlying DDS objects did get deleted as a consequence of deleting the participant).

I've added deleting them to the `rmw_context_impl_t` destructor and adjusted the initialisation code a bit so that I think it should all work fine now. I've tried it, also with some fault injection, but I might be getting some of the C++-ey bits wrong. @rotu perhaps you could have a look as well?